pkg/metallb/frr.go

```diff
@@ -0,0 +1,675 @@
+	builder.Definition.Spec.BGP.Routers[routerIndex].Neighbors[neighborIndex].ToReceive.Allowed.Mode = frrtypes.AllowAll
+
+	return builder
+}
+
+// WithToReceiveModeFiltered allows only specified prefixes to be received.
+func (builder *FrrConfigurationBuilder) WithToReceiveModeFiltered(prefixes []string,
+	routerIndex, neighborIndex int) *FrrConfigurationBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof(
+		"Allows the FRR nodes to receive specific routes in namespace %s in mode type: filtered",
+		builder.Definition.Namespace)
+
+	// Check if BGP routers exist
+	if len(builder.Definition.Spec.BGP.Routers) == 0 {
+		glog.V(100).Infof("Invalid routerIndex: %d no router was found", routerIndex)
+
+		builder.errorMsg = "No BGP routers defined in WithToReceiveModeFiltered."
+	}
+
+	// Check if the neighborIndex is within bounds
+	if neighborIndex < 0 || neighborIndex >= len(builder.Definition.Spec.BGP.Routers[routerIndex].Neighbors) {
+		glog.V(100).Infof("Invalid neighborIndex: %d", neighborIndex)
+
+		builder.errorMsg = fmt.Sprintf("Invalid neighborIndex: %d", neighborIndex)
+	}
+
+	// Validate CIDR prefixes
+	for _, prefix := range prefixes {
+		if _, _, err := net.ParseCIDR(prefix); err != nil {
+			glog.V(100).Infof("the frrConfiguration prefix %s is not a valid CIDR", prefix)
+			builder.errorMsg = fmt.Sprintf("The prefix %s is not a valid CIDR", prefix)
+		}
+	}
+
+	if builder.errorMsg != "" {
+		return builder
+	}
+
+	// Prepare the list of allowed prefixes
+	allowedPrefixes := make([]frrtypes.PrefixSelector, len(prefixes))
+
+	for i, prefix := range prefixes {
+		glog.V(100).Infof("AllowMode 'filtered' receives only routes within the prefix %v", prefix)
+		allowedPrefixes[i] = frrtypes.PrefixSelector{Prefix: prefix}
+	}
+
+	// Update the specific neighbor's ToReceive field
+	builder.Definition.Spec.BGP.Routers[routerIndex].Neighbors[neighborIndex].ToReceive.Allowed.Prefixes = allowedPrefixes
+
+	return builder
+}
+
+// WithBGPNeighbor defines a single neighbor IP and ASN number.
+func (builder *FrrConfigurationBuilder) WithBGPNeighbor(bgpPeerIP string,
+	remoteAS uint32, routerIndex int) *FrrConfigurationBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof(
+		"Define BGP Neighbor %s with peer IP %s and ASN %v",
+		builder.Definition.Name, bgpPeerIP, remoteAS)
+
+	if net.ParseIP(bgpPeerIP) == nil {
+		glog.V(100).Infof("The peerIP for the frrConfiguration bgp peer contains invalid ip address %s",
+			bgpPeerIP)
+
+		builder.errorMsg = "frrConfiguration 'peerIP' of the BGPPeer contains invalid ip address"
+	}
+
+	if remoteAS < 1 || remoteAS > 65535 {
+		glog.V(100).Infof("The remote ASN of the frrConfiguration must be between 1 and 65535")
+
+		builder.errorMsg = "frrConfiguration 'remote AS' must be between 1 and 65535"
+	}
+
+	// Check if BGP routers exist
+	if len(builder.Definition.Spec.BGP.Routers) == 0 {
```

Comment:
if routerIndex -1 ? try to validate that the router with [routerIndex] exists.
related to all the functions

Reply:
Ok it will probably panic. I changed the logic.